internal-packages/next-dev/src/index.ts

```diff
@@ -0,0 +1,204 @@
+	 * (so that the same data can be easily used by both the next dev server and `wrangler pages dev`).
+	 * If `false` is specified no data is persisted on the filesystem.
+	 */
+	persist?: false | string;
+};
+
+/**
+ * Creates the miniflare instance that we use under the hood to provide access to bindings.
+ *
+ * @param options the user provided options
+ * @returns the new miniflare instance.
+ */
+async function instantiateMiniflare(
+	options: DevBindingsOptions,
+): Promise<Miniflare> {
+	const { workerOptions, durableObjects } =
+		(await getDOBindingInfo(options.durableObjects)) ?? {};
+
+	const { kvNamespaces, r2Buckets, d1Databases, services, textBindings } =
+		options;
+	const bindings = {
+		bindings: textBindings,
+		kvNamespaces,
+		durableObjects,
+		r2Buckets,
+		d1Databases,
+		services,
+	};
+
+	const serviceBindings = await getServiceBindings(services);
+
+	const workers: WorkerOptions[] = [
+		{
+			...bindings,
+			modules: true,
+			script: '',
+			serviceBindings,
+		},
+		...(workerOptions ? [workerOptions] : []),
+	];
+
+	// we let the user define where to persist the data, we default back
+	// to .wrangler/state/v3 which is the currently used wrangler path
+	// (this is so that when they switch to wrangler pages dev they can
+	// still interact with the same data)
+	const persist = options?.persist ?? '.wrangler/state/v3';
+
+	const mf = new Miniflare({
+		workers,
+		...(persist === false
+			? {
+					// the user specifically requested no data persistence
+			  }
+			: {
+					kvPersist: `${persist}/kv`,
+					durableObjectsPersist: `${persist}/do`,
+					r2Persist: `${persist}/r2`,
+					d1Persist: `${persist}/d1`,
+			  }),
+	});
+
+	return mf;
+}
+
+/**
+ * Next.js uses the Node.js vm module's `runInContext()` function to evaluate the edge functions
+ * in a runtime context that tries to simulate as accurately as possible the actual production runtime
+ * behavior, see: https://github.com/vercel/next.js/blob/ec0a8da/packages/next/src/server/web/sandbox/context.ts#L450-L452
+ *
+ * This function monkey-patches the Node.js vm module to override the `runInContext()` function so that
+ * miniflare binding proxies can be added to the runtime context's `process.env` before the actual edge
+ * functions are evaluated.
+ *
+ * @param bindings array containing the miniflare binding proxies to add to the runtime context's `process.env`
+ */
+function monkeyPatchVmModule(bindings: Record<string, unknown>) {
+	// eslint-disable-next-line @typescript-eslint/no-var-requires
+	const vmModule = require('vm');
+
+	const originalRunInContext = vmModule.runInContext.bind(vmModule);
+
+	vmModule.runInContext = (
+		...args: [
+			string,
+			Record<string, unknown> & { process?: { env?: Record<string, unknown> } },
+			...[unknown],
+		]
+	) => {
+		const runtimeContext = args[1];
+		if (
+			runtimeContext.process?.env &&
+			!runtimeContext.process?.env?.['BINDINGS_PROXY_SET']
```

Comment:
Do you still need the optional chaining operators here?

Could it be:

```suggestion
			runtimeContext.process?.env &&
			!runtimeContext.process.env['BINDINGS_PROXY_SET']
```

Reply:
Is there any concern over a conflict with user code having an actual env var called `BINDINGS_PROXY_SET`?

Reply:
ah! yeah that's a totally valid point! I should be able to just use a unique symbol here... let me have a look! 🙂👍

Reply:
I've applied the code to use a symbol, so there shouldn't be any concern anymore 🙂

thanks for pointing this out 😄